OpenRA.Mods.Common/Traits/Upgrades/ProximityExternalCondition.cs

```diff
@@ -129,26 +131,24 @@ public void UnitProducedByOther(Actor self, Actor producer, Actor produced)
 					return;
 
 				var um = produced.TraitOrDefault<UpgradeManager>();
-				if (um != null)
-					foreach (var u in info.Upgrades)
-						if (um.AcknowledgesUpgrade(produced, u))
-							um.GrantTimedUpgrade(produced, u, 1);
+				if (um != null && um.AcceptsExternalCondition(produced, info.Condition))
```

Comment:
I assume this is intentional, but I'm struggling to get my head around why this doesn't need a `ContainsKey` check, could you elaborate a bit?